packages/jupyterlab-manager/src/plugin.ts

```diff
@@ -1,101 +1,189 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 
+
+import { ISettingRegistry } from '@jupyterlab/coreutils';
+
 import {
   DocumentRegistry
 } from '@jupyterlab/docregistry';
 
 import {
-  INotebookModel
-} from '@jupyterlab/notebook';
-
-import {
-  NotebookPanel
+  INotebookModel, INotebookTracker, Notebook
 } from '@jupyterlab/notebook';
 
 import {
   JupyterFrontEndPlugin, JupyterFrontEnd
 } from '@jupyterlab/application';
 
 import {
-  IDisposable, DisposableDelegate
+  IMainMenu,
+} from '@jupyterlab/mainmenu';
+
+import {
+  RenderMimeRegistry, IRenderMimeRegistry
+} from '@jupyterlab/rendermime';
+
+import {
+  CodeCell
+} from '@jupyterlab/cells';
+
+import {
+  each
+} from '@phosphor/algorithm';
+
+import {
+  DisposableDelegate
 } from '@phosphor/disposable';
 
+import {
+  AttachedProperty
+} from '@phosphor/properties';
+
 import {
   WidgetRenderer
 } from './renderer';
 
 import {
-  WidgetManager
+  WidgetManager, WIDGET_VIEW_MIMETYPE
 } from './manager';
 
 import {
   OutputModel, OutputView, OUTPUT_WIDGET_VERSION
 } from './output';
 
 import * as base from '@jupyter-widgets/base';
+
+// We import only the version from the specific module in controls so that the
+// controls code can be split and dynamically loaded in webpack.
 import {
   JUPYTER_CONTROLS_VERSION
 } from '@jupyter-widgets/controls/lib/version';
 
 import '@jupyter-widgets/base/css/index.css';
 import '@jupyter-widgets/controls/css/widgets-base.css';
 
-const WIDGET_MIMETYPE = 'application/vnd.jupyter.widget-view+json';
+const WIDGET_REGISTRY: base.IWidgetRegistryData[] = [];
+
+/**
+ * The cached settings.
+ */
+const SETTINGS: {saveState: boolean} = {saveState: false};
 
 export
-type INBWidgetExtension = DocumentRegistry.IWidgetExtension<NotebookPanel, INotebookModel>;
+function registerWidgetManager(nb: Notebook, context: DocumentRegistry.IContext<INotebookModel>, rendermime: RenderMimeRegistry) {
+  let wManager = Private.widgetManagerProperty.get(context);
+  if (!wManager) {
+    wManager = new WidgetManager(context, rendermime, SETTINGS);
+    WIDGET_REGISTRY.forEach(data => wManager.register(data));
+    Private.widgetManagerProperty.set(context, wManager);
+  }
 
+  // For any widgets that have already been rendered with the placeholder, set
+  // the renderer. This iteration structure is closely tied to the structure of
+  // the notebook widget.
+  nb.widgets.forEach(cell => {
```

Comment:
I just pushed a commit that (a) refactors the awkward logic into a self-contained iterator, and (b) attempts to also iterate through linked views, using even *more* awkward logic that relies on specific jlab details. It's hard for me to test, though, since linked output views aren't restored in the latest jlab release.

Reply:
(note that both of these awkward iterators are just an interim solution until there might be a more supported clean way of iterating through renderers in jlab itself)